### ISSUE TITLE
Performance issues when opening large cbz files

### ISSUE THREAD

**Issue by moshiur-raj** [johnfactotum/foliate#813]
**Describe the bug**
When opening large cbz files (~900MB) foliate uses too much ram. At startup it consumes all of my ram (I have 8GB of ram), the system just hangs up and after a while the ram usage stays at ~2GB (most of the memory is consumed by the webkit network process). Meanwhile evince uses ~70MB of ram for the same cbz file. Foliate also takes longer to load the file.

**To Reproduce**
1. Open a large cbz file with foliate.
2. Monitor the system using system monitor.

**Expected behavior**
Foliate should not use this much ram.

**Version:**
 - Foliate version: 2.6.3
 - OS/Distribution and version: Arch-LInux
 - Desktop environment: Gnome 41.1
 - Installation method: package from official arch repo

**Comment by johnfactotum** [comment]
It's the inherent limitation of our current method of handling comic book files. It's all handled in the WebView, so it has to read the whole file and unarchive it in memory before it can even render a single image.

And there's no way around this when using Epub.js; no matter how the file is processed, ultimately it needs to be completely unarchived in memory if it is to be consumed and displayed by Epub.js.

This is the case for EPUB files and all other formats as well, though it's usually less noticeable, as they tend to be smaller in size.

So the only solution is to stop using Epub.js, at least for comic book files.

**Comment by moshiur-raj** [comment]
I see. The thing is most cbz files are large. This makes foliate unsuitable for reading cbz files. But I like using foliate instead of evince for cbz files because of it's minimal ui and wonderful design. Evince uses poppler for rendering, right? Are there any plans for using that or some other software?

**Comment by johnfactotum** [comment]
>Evince uses poppler for rendering, right? Are there any plans for using that or some other software?

I think it only uses Poppler for PDFs. It uses libarchive to unzip CBZs and draws the images directly with (I presume) Cairo, or something like that.

Anyway, we could use Evince as a library and use its widget to render CBZs; or we could implement our own (which would be much easier than implementing our own viewer for PDFs). Both approaches has some pros and cons.

The main obstacle of the latter option, though, is the lack of a GObject-Introspection-capable archive library, and I don't want to have to rewrite Foliate to use non-GJS. So it's a bit of a hassle. Either way, I probably won't be able to address this issue any time soon.

**Comment by johnfactotum** [comment]
And another thing one might look into is replacing JSZip, which I believe isn't very performant compared to some newer alternatives, which could improve performance for all zipped files.

**Comment by johnfactotum** [comment]
I thought a bit more about this. One idea is that we could read the archived file outside the WebView, and set up a local server. Then, in the WebView, Epub.js can consume the file as an unarchived directory through the server. This would mean that it no longer needs to unarchive the whole file in memory; instead only the parts requested by Epub.js would be unarchived and returned from the server. This even has the side effect of increasing security, as the WebView would no longer require unrestricted filesystem access (though this is less of a problem in Flatpak).

**Comment by itprojects** [comment]
> read the archived file outside the WebView, and set up a local server

This would be a major change in the infrastructure.
If you're going to split the back-end, why do it _just_ for cbz?

>  even has the side effect of increasing security, as the WebView would no longer require unrestricted filesystem access (though this is less of a problem in Flatpak).

It won't necessarily be more secure; there will be more elements, and configurations to test and consider.

**Comment by moshiur-raj** [comment]
I think implementing lazy loading is a good idea. It will make loading larger files (not just cbz) a lot snappier. Currently foliate is not a good choice if your filesize is large, it consumes too much ram.

**Comment by johnfactotum** [comment]
> This would be a major change in the infrastructure. If you're going to split the back-end, why do it _just_ for cbz?

It's not really as big a change as one might think. The structure can be used for all files, but it's particularly easy to implement for EPUB files, as it's just a matter of setting up the server and let Epub.js consume it. It's a bit more involved for non-EPUB files, as their conversion to EPUB currently happens inside the WebView.

> It won't necessarily be more secure; there will be more elements, and configurations to test and consider.

It would fix a known problem where the book can access any local file, which is the main reason why Foliate currently forbids all external resources with no option to enable them.

It would also mean that we no longer need to bundle JSZip, libarchivejs, and crypto-js. All these functions will live in the server process, which if we use Python, can readily use its standard crypto and zip libraries instead. As for libarchivejs, we can just use ctypes to access it in Python. So this actually _reduces_ elements.

And this makes it easier to make more transparent and reproducible builds, as we now only need to build Epub.js, which currently we don't ship it minified, anyway. So this will help with Debian packaging as well.

**Comment by itprojects** [comment]
@johnfactotum  Will this re-designed Foliate have to unarchive a file on disk? (privacy context)

`cbz/cbr` formats are known for their large size, delivering on a faithful comic-book experience. Comic book software automatically focuses, enlarges, animates parts of the page, when the 'next' button is pressed. This feature is not typical of ebooks (and shouldn't be), but without it Foliate will be a much inferior to any dedicated comic book reader.

> It's not really as big a change as one might think.

Adding a server, more python, and ctypes - why not rewrite in Foliate in Vala (or Python)? (for memory management/threads)

Last, but not least: Foliate _still_ works.

**Comment by johnfactotum** [comment]
> Will this re-designed Foliate have to unarchive a file on disk?

No, it won't. It will unarchive files in memory.

> Adding a server, more python, and ctypes - why not rewrite in Foliate in Vala (or Python)?

Using GJS still has many advantages, and even though Foliate 3.0 is going to be mostly rewritten, it is certainly still going to reuse a large amount of code from 2.0.

The main disadvantage of GJS is the fact the lack of libraries, in particular the single most important library when it comes to e-books, an unarchiver, which we can work around by adding a small Python script.

How small? If we only do it for EPUB files, it can probably be done with maybe 50 lines of code or fewer. It's just so small a change and so many advantages:
- It no longer unarchives the whole file in RAM, significantly reducing memory usage.
- JSZip, as I've mentioned, is not very performant. So even if we disregard memory issues, switching away from JSZip would probably make things faster.
- It no longer requires unlimited filesystem access in the WebView, which is a _huge_ security risk. Please see https://github.com/johnfactotum/foliate/issues/391#issuecomment-637124746. There are other, much, much more involved ways to fix this bug, but it's a nice side effect that comes with almost no cost.
- It can use Gio to read remote files. I believe this would fix the bug where you can't open files from, e.g., Samba. There are also other ways to fix this bug, of course, but it's another nice side effect.
- As mentioned above, it would greatly reduce the amount of bundled JS code, so reducing package size, and also helping with distro packaging.

There are just so many advantages and I can't really think of a single drawback except maybe some slight overhead with running a Python process, which is trivial compared to the benefits it brings.

For non-EPUB files, this new structure would require a bit more work. But comic book files are pretty much trivial to convert to EPUB, so rewriting the conversion in Python is trivial. As for FB2, you would need to load the whole file into memory, anyway, so it would be the main "victim" of this change, as it doesn't benefit from it and requires the most work to implement. We could probably keep FB2 conversion in the WebView, as I do not want to rewrite it again, and just pass things around a bit. So it actually seems pretty trivial as well.

**Comment by itprojects** [comment]
Your comments show that you've researched this fully.

PDF's was mentioned earlier; any news? (for local server Foliate variation)

**Comment by johnfactotum** [comment]
Not really. If we do use Python then maybe we can use PyMuPDF but I haven't looked much into it.

**Comment by itprojects** [comment]
PyMuPDF should be great. It's small and simple.

Maybe it can work for cbz/cbr?

**Comment by johnfactotum** [comment]
There isn't any advantage in using mupdf for CBZs as (1) CBZ is a very simple format to handle directly without any special libraries, (2) mupdf doesn't support other comic book formats.

Also, even for PDF files, it is doubtful whether using mupdf would be any better than poppler, which we can use directly in GJS.

So please stop conflating this issue with the issue about PDF support. PDFs are much, much more complicated and totally different from all the other formats that Foliate currently supports. The one format that does have something in common with PDF is DjVu, which mupdf doesn't support.

TL;DR: there's no real advantage in using mupdf, and at any rate it is bears no relation to this issue whatsoever.

---

To be clear: this issue is about increasing performance by not reading the whole file into memory when only parts of the file is needed (i.e. it needs to do seeking when reading the file). There are perhaps other ways to solve this issue, but what I've stated is what I believe to be the easiest solution. It may not be the best solution but it's the easiest to do. It's just a small workaround to fix a specific performance issue. That's all.

**Comment by johnfactotum** [comment]
Fixed in the [`gtk4`](https://github.com/johnfactotum/foliate/tree/gtk4) branch.
